cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+
+That directory needs to contain 3 files.
+
+* cbmc-proof.txt
+* Makefile
+* XXX_harness.c
+
+where "XXX" is the name of the function being proved - same as the directory name.
+
+I suggest that you copy these files from an existing proof directory and modify the latter two. The `cbmc-proof.txt` file is just a marker that says "this directory contains a CBMC proof" to the tools, so no modification is required.
+
+### Update Makefile
+
+The `Makefile` sets options and targets for this proof. Let's imagine that the function we want to prove is called `XXX` (without the `$(KYBER_NAMESPACE)` prefix).
+
+Edit the Makefile and update the definition of the following variables:
+
+* HARNESS_FILE - should be `XXX_harness`
+* PROOF_UID - should be `XXX`
+* PROJECT_SOURCES - should the files containing the source code of XXX
+* CHECK_FUNCTION_CONTRACTS - set to the `XXX`, but _with_ the `$(KYBER_NAMESPACE)` prefix if required
+* USE_FUNCTION_CONTRACTS - a list of functions that `XXX` calls where you want CBMC to use the contracts of the called function for proof, rather than 'inlining' the called function for proof. Include the `$(KYBER_NAMESPACE)` prefix if required
+* CBMCFLAGS - additional flags to pass to the final run of CBMC. This is normally set to `--smt2` which tells cbmc to run Z3 as its underlying solver. Can also be set to `--bitwuzla` which is sometimes better at generaing counter-examples when Z3 fails.
+* FUNCTION_NAME - set to `XXX` with the `$(KYBER_NAMESPACE)` prefix if required
+* CBMC_OBJECT_BITS. Normally set to 8, but might need to be increased if CBMC runs out of memory for this proof.
+
+For documentation of these (and the other) options, see the `cbmc/proofs/Makefile.common` file.
+
+### Update harness function
+
+The file `XXX_harness.c` should declare a single function called `XXX_harness()` that calls `XXX` exactly one, with appropriately typed parameters. Using contracts, this harness function should NOT need to contain any CBMC `ASSUME` or `ASSERT` statements at all.
+
+# Supply top-level contracts
+
+Now for the tricky bit. Does `XXX` need any top-level pre- and post-condition contracts, other that those inferred by its type-signature?
+
+A common case is where a function has an array parameter, where the numeric range of the array's elements are NOT just the range of the underlying predefined type.
+
+Secondly, a parameter of a predefined integer type might have to be constrained to a small range.
+
+Check for any BOUND macros in the body that give range constraints on the parmeters, and make sure that your contracts are identical or consistent with those. Also read the comments for XXX to make sure they agree with your contracts.
+
+The order of the top-level contracts for a function is always:
+```
+t1 XXX(params...)
+    // clang-format off
+REQUIRES()
+ASSIGNS()
+ENSURES();
+// clang-format on
+```
+with a final semi-colon on the end of the last one.
```

Comment:
Again, this is quite unexpected, so maybe worth emphasizing more.

Reply:
Added extra sentence to emphasize this rule.